### ISSUE TITLE
sokol_app in custom iOS app

### ISSUE THREAD

**Issue by PeterHiber** [floooh/sokol#520]
Hi and thanks for `sokol_app` and `sokol_gfx`! They are great and have been very useful for us in the development of our new cross-platform game engine. :)

Currently we have a bit of an interesting problem, so we are mainly creating this issue to get your thoughts.

Basically, we have decided to integrate the new game engine into a custom app (built using react native) on iOS. The react app and engine are not running at the same time, but we are switching between them. At first we thought we could solve it by using `SOKOL_NO_ENTRY` and then launch the engine from the react app, but we ran into some problems because both sokol_app and react native have their own `AppDelegate`'s which clashes with each other.

At this point I think our options are to either drop sokol_app (which would be a shame because it's great) or modify sokol_app to support this use case.

Specifically we have been looking at not using the `_sapp_ios_run()`, instead using a shared App Delegate. This would mean that we'd setup everything needed for our game engine and switching the root view when starting the game in that App Delegate, basically what's in the `didFinishLaunchingWithOptions` of `_sapp_app_delegate`, and then go back to the react native part when leaving the game (switching back to that view).

So this became a bit of a wall of text, so basically our questions are:

* Do you think this approach (alternate iOS run function that takes an existing AppDelegate) is sensible?
* Do you have another idea that you think would be simpler?
* Is there any functionality in `sokol_app` that we have missed that might make this easier?
* Would you be interested in having this upstreamed if we did this work? We would love to not have to fork `sokol_app` ;)

**Comment by floooh** [comment]
Hmm, I can't think immediately of a "good" solution. Would it really be enough to just pass another application delegate to the run-function? Because _sapp_ios_run() is still like its own application entry point that doesn't return control to the caller, I would expect that this clashes with the event loop of the "wrapper application". If this does however work, than we could simply add a new ios-specific item to the sapp_desc struct to either pass an application delegate pointer or class name, and if it actually works with so few changes that I have nothing against merging this to master. I've got a feeling though that it probably won't be that easy ;)

**Comment by PeterHiber** [comment]
Hi again!

We have come a bit further with this and we think we might have an approach that could for us and might be clean enough to upstream. Not done yet, but we thought it might be good to ask what you think so far. :)

Basically, we want to create our own AppDelegate and use instead of the one inside `sokol_app`. As a first step we decided to make everything in sokol_app public (`#define _SOKOL_PRIVATE`) to make things easier to test. Then we copied the contents of `_sapp_ios_run()` directly to our own `main()`, worked perfectly of course.

Next step we copied sokol_apps entire `_sapp_app_delegate`, renamed it and placed it in our main file and attempted to use that instead. This of course does not work because we can't access `static _sapp_t _sapp;` in a good way from outside sokol_app.

To solve above we added a small layer of indirection inside sokol_app:

Before:
```
@implementation _sapp_app_delegate

// ...

- (void)applicationWillResignActive:(UIApplication *)application {
        if (!_sapp.ios.suspended) {
            _sapp.ios.suspended = true;
            _sapp_ios_app_event(SAPP_EVENTTYPE_SUSPENDED);
       }
}

// ...

@end
```

After:
```
_SOKOL_PRIVATE void _sapp_app_delegate_applicationWillResignActive(UIApplication* application)
{
    if (!_sapp.ios.suspended) {
        _sapp.ios.suspended = true;
        _sapp_ios_app_event(SAPP_EVENTTYPE_SUSPENDED);
    }
}

@implementation _sapp_app_delegate

// ...

- (void)applicationWillResignActive:(UIApplication *)application {
    _sapp_app_delegate_applicationWillResignActive(application);
}

// ...

@end
```

This way it becomes trivial for us to create our own AppDelegate outside SokolApp, our complete main:

```
#include "sokol_app.h"

#import <UIKit/UIKit.h>

void _sapp_init_state(const sapp_desc* desc);
bool _sapp_app_delegate_didFinishLaunchingWithOptions(NSDictionary* launchOptions);
void _sapp_app_delegate_applicationWillResignActive(UIApplication* application);
void _sapp_app_delegate_applicationDidBecomeActive(UIApplication* application);
void _sapp_app_delegate_applicationWillTerminate(UIApplication* application);

@interface _sapp_app_delegate2 : NSObject<UIApplicationDelegate>
@end

@implementation _sapp_app_delegate2
- (BOOL)application:(UIApplication*)application didFinishLaunchingWithOptions:(NSDictionary*)launchOptions {
    return _sapp_app_delegate_didFinishLaunchingWithOptions(launchOptions);
}

- (void)applicationWillResignActive:(UIApplication *)application {
    _sapp_app_delegate_applicationWillResignActive(application);
}

- (void)applicationDidBecomeActive:(UIApplication *)application {
    _sapp_app_delegate_applicationDidBecomeActive(application);
}

- (void)applicationWillTerminate:(UIApplication *)application {
    _sapp_app_delegate_applicationWillTerminate(application);
}
@end


int main(int argc, char* argv[]) {
     _sapp_init_state(&(sapp_desc){
        .init_cb = hbrInit,
        .frame_cb = hbrFrame,
        .cleanup_cb = hbrCleanup,
        .event_cb = NULL,
        .width = 400,
        .height = 300,
        .gl_force_gles2 = true,
        .window_title = "Minimal iOS Test",
    });
    UIApplicationMain(argc, argv, nil, NSStringFromClass([_sapp_app_delegate2 class]));
}

```

So basically TL;DR:

* Add a layer of abstraction to sokol_app's AppDelegate, make this abstraction exposable.
* Add some way of specifying your own AppDelegate when initializing sokol_app (not done yet)


Might add that we still haven't gotten to the switching between the game engine and react native yet, so above might not be enough. 😅
